mlflow/pyfunc/__init__.py

```diff
@@ -370,62 +370,149 @@ def all_ints(xs):
         )
 
 
-def _enforce_schema(pdf: PyFuncInput, input_schema: Schema):
+def _enforce_tensor_spec(values: np.ndarray, tensor_spec: TensorSpec):
     """
-    Enforce column names and types match the input schema.
+    Enforce the input tensor shape and type matches the declared in model input schema.
+    """
+    expected_shape = tensor_spec.shape
+    actual_shape = values.shape
+    if len(expected_shape) != len(actual_shape):
```

Comment:
Suppose the stored spec is (-1,28,28,3) and a user passes a single input of (28,28,3), I think we should allow that input.
It will fail the check here.

Reply:
cc @tomasatdatabricks

Reply:
hm, i think it makes more sense to always expect input to be a batch (i believe tf and other frameworks i've used in the past have this expectation too)

Reply:
I think that makes sense. 
What about the other way, the stored spec does not contain a batch (28, 28, 3) and the input contains a batch?

Reply:
I think we should always honor the expected shape as the DL frameworks do.